lib/zoho_resource_client.rb

```diff
@@ -0,0 +1,73 @@
+class ZohoResourceClient
+  ZOHO_ORG_ID = ArchiveConfig.ZOHO_ORG_ID
+  CONTACT_SEARCH_ENDPOINT = 'https://desk.zoho.com/api/v1/contacts/search'.freeze
+  CONTACT_CREATE_ENDPOINT = 'https://desk.zoho.com/api/v1/contacts'.freeze
+  TICKET_CREATE_ENDPOINT = 'https://desk.zoho.com/api/v1/tickets'.freeze
+
+  def initialize(access_token:, email:, username: nil)
+    @access_token = access_token
+    @email = email
+    @username = username
+  end
+
+  def retrieve_contact_id
+    (find_contact || create_contact).fetch('id')
+  end
+
+  def create_ticket(ticket_attributes:)
+    response_raw = HTTParty.post(
+      TICKET_CREATE_ENDPOINT,
+      headers: headers,
+      body: ticket_attributes.to_json
+    )
+    JSON.parse(response_raw.body)
+  end
+
+  private
+
+  def find_contact
+    response_raw = HTTParty.get(
+      CONTACT_SEARCH_ENDPOINT + '?' + query_string(search_params),
+      headers: headers,
+    )
+    return if response_raw.nil?
+
+    JSON.parse(response_raw.body).fetch('data').first
+  end
+
+  def create_contact
+    response_raw = HTTParty.post(
+      CONTACT_CREATE_ENDPOINT,
+      headers: headers,
+      body: contact_body.to_json
+    )
+    JSON.parse(response_raw.body)
+  end
+
+  def search_params
+    {
+      email: @email,
+      limit: 1,
+      sortBy: 'modifiedTime',
+    }
+  end
+
+  def headers
+    {
+      "Content-Type" => "application/json",
+      "orgId" => ZOHO_ORG_ID,
+      "Authorization" => "Zoho-oauthtoken #{@access_token}"
+    }
+  end
+
+  def contact_body
+    {
+      "lastName" => @username.presence || @email,
```

Comment:
Correct me if I'm wrong, but it looks like `@username` is only used here, which means that it's recorded when the user submits their first ticket, but never updated in subsequent tickets. So if a user changes their name between tickets (maybe they're logged-in and changed their username? maybe they made a typo the first time?), the second ticket will not contain the new name.

Maybe the `lastName` in the contact should be updated if it doesn't match the name the user submitted, so that Support/Abuse always knows the current preferred name of the person they're interacting with? Or maybe the `@username` should be added as a custom field, so that each ticket has the name that the user wanted to be associated with that ticket? (In that case, it might make sense to always set `lastName` equal to the email, so that it never needs to be updated.)

Reply:
I like the idea of always using the email as the lastName and making username a custom field.

Reply:
Do you want to liaise with Support to decide what they want the custom field to be named? `username` is a bit misleading, given that it's not pulled from the Archive username.

Reply:
Can we just go with "Name" for the custom field (the forms both say "Your name" so it would be most logical)? If not, then yeah, I'll talk to Support and PAC.

Reply:
"Name" it is!